cpp/src_prims/cache/cache.h

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_utils.h>
+#include <cub/cub.cuh>
+#include "cache_util.h"
+#include "common/cumlHandle.hpp"
+#include "common/device_buffer.hpp"
+#include "cuML.hpp"
+#include "ml_utils.h"
+
+namespace MLCommon {
+namespace Cache {
+
+using namespace MLCommon;
+
+/**
+* @brief Associative cache with least recently used replacement policy.
+*
+* SW managed cache in device memory, for ML algos where we can trade memory
+* access for computation. The two main functions of this class are the
+* management of cache indices, and methods to retreive/store data using the
+* cache indices.
+*
+* The index management can be considered as a hash map<int, int>, where the int
+* keys are the original vector indices that we want to store, and the values are
+* the cache location of these vectors. The indices are hashed into a bucket
+* whose size equals the associativity. These are the cache sets. If a cache
+* set is full, then new indices are stored by replacing the oldest entries.
+*
+* Using this index mapping we implement methods to store and retrive data from
+* the cache buffer, where a unit of data that we are storing is math_t[n_vec].
+* For example in SVM we store full columns of the kernel matrix at each cache
+* entry.
+*
```

Comment:
Minor knitpick: adding a usage example here in the documentation would be very useful, just like done in this header file: https://github.com/rapidsai/cuml/blob/branch-0.9/cpp/src_prims/common/device_buffer.hpp

Reply:
That is a very good point, I have added an example.